### PR TITLE
Introduces PipelineValidator that checks the well-formedness of a pipeline

### PR DESCRIPTION
**Pull request by jkff**
Motivation: being more confident about various code that manipulates the pipeline, e.g. fusion code, SDF expansion, etc.

Didn't add explicit tests for the validation code - it will be implicitly tested by every time every portable pipeline runs; the incremental benefit of adding explicit tests doesn't seem to justify the cost. Any false positives will be caught by a respective pipeline, false negatives will be caught at runtime.

(piggybacked here: deleted unused and empty ReferenceRunner class - the proper name for that concept is PortableRunner. At some point we should rename the module to beam-runners-portable-java too)

R: @lukecwik

### CODE REVIEW

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/PipelineValidator.java [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+import org.apache.beam.model.pipeline.v1.RunnerApi.ParDoPayload;
+import org.apache.beam.model.pipeline.v1.RunnerApi.StandardPTransforms.CombineComponents;
+import org.apache.beam.model.pipeline.v1.RunnerApi.StandardPTransforms.Composites;
+import org.apache.beam.model.pipeline.v1.RunnerApi.StandardPTransforms.Primitives;
+import org.apache.beam.model.pipeline.v1.RunnerApi.StandardPTransforms.SplittableParDoComponents;
+import org.apache.beam.model.pipeline.v1.RunnerApi.TestStreamPayload;
+import org.apache.beam.model.pipeline.v1.RunnerApi.WindowIntoPayload;
+import org.apache.beam.model.pipeline.v1.RunnerApi.WindowingStrategy;
+
+public class PipelineValidator {
```
Comment (lukecwik) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/PipelineValidator.java]: class comment

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/PipelineValidator.java [hunk 2]
```diff
@@ -0,0 +1,267 @@
+package org.apache.beam.runners.core.construction.graph;
```
Comment (lukecwik) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/PipelineValidator.java]: ASF license

**Comment by jkff** [comment]
Hm Jenkins seems to not be picking up this change. Guess I'll have to wait a bit more.

**Comment by jkff** [comment]
retest this please

**Comment by jkff** [comment]
retest this please

**Pull request merged**